File: python/p_384.py
"""
Shuffle a set of numbers without duplicates
"""

class Solution(object):

    def __init__(self, nums):
        """
        
        :type nums: List[int]
        :type size: int
        """
        self.nums = nums
        

    def reset(self):
        """
        Resets the array to its original configuration and return it.
        :rtype: List[int]
        """
        return self.nums
        

    def shuffle(self):
        """
        Returns a random shuffling of the array.
        :rtype: List[int]
        """
        nums2 = self.nums[:]
        random.shuffle(nums2)
        return nums2
        


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()

File: cpp/temp/p_384.cpp
/**
 * rand() is defined in stdlib.h, which generates a random int, unlike in Python or JS
 * where a float between 0 and 1 is generated
 * std::swap can be used to swap the value of two variable 
 */
#include <vector>
#include <iostream>
#include "stdlib.h"
using namespace std;

class Solution {
    vector<int> nums;
public:
    Solution(vector<int> nums) {
        this->nums = nums;
    }
    
    /** Resets the array to its original configuration and return it. */
    vector<int> reset() {
        return nums;
    }
    
    /** Returns a random shuffling of the array. */
    vector<int> shuffle() {
        vector<int> nums2(nums);
        int n = int(nums.size());
        
        for(int i = n-1; i >= 0; --i) {
            int j = rand() % (i+1);
            swap(nums2[i], nums2[j]);
        }
        return nums2;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * vector<int> param_1 = obj.reset();
 * vector<int> param_2 = obj.shuffle();
 */
